nixos/testing: factor out VM functionality

mkenigs opened the pull request:
###### Motivation for this change
- The first commit is just a cleanup and should be easy to merge.

- The second doesn't change any functionality, but it's a step towards allowing
Nixos tests to be run in containers. Splitting out VM functionality from
Machine functionality will allow easily creating a Container subclass
for running tests in containers.

  - The external interface of test-driver is changed from create_machine()
to VM().

- The third commit has the only real code changes I had to make. Not sure why but I had to add some types and assertions for tests to pass. I can squash it if it looks good

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mkenigs commented:
@edolstra I think this is the next step towards tests in containers after https://github.com/NixOS/nix/pull/3600/ gets merged

tfc commented:
whoa, where does this suddenly come from... i think we have a little clash here.

@blaggacao is in the progress of a similar thing: https://github.com/NixOS/nixpkgs/pull/126713

@mkenigs can we maybe have a chat and talk about what your plan is in order to converge?

mkenigs commented:
@tfc pinged you on Element

tfc commented:
> @tfc pinged you on Element

plz ping @jonge:ukvly.org, i don't really use the other account

mkenigs committed:
Remove unused allow_reboot instance variable

mkenigs committed:
Remove unused keep_vm_state instance variable

mkenigs committed:
(squash) add class structure

mkenigs committed:
(squash) split up __init__ and instance variables

mkenigs committed:
(squash) add abstract methods

mkenigs committed:
(squash) move methods into appropriate classes

mkenigs committed:
(squash) move sleep()

sleep() could be put in ExecuteMixin, but doing so would require adding
an abstract method to Machine. Since sleep() is only called once in
PrivateVM, just move it there

mkenigs committed:
(squash) move all remaining methods

mkenigs committed:
(squash) replace Machine with VM

mkenigs commented:
@tfc completely reworked everything, let me know what you think!

Ericson2314 commented:
@Ma27 would you be interested in shepherding this?

mkenigs commented:
To summarize some conversation I had a while back with @tfc: this is mostly good to go, but he didn't want to merge it until we have a better idea of what requirements will be for a container subclass. Using containers is related to https://github.com/NixOS/nix/pull/3600/

Ericson2314 commented:
It does seem like a chicken and egg problem, but the sort I think @Ma27 is uniquely well positioned to untangle :).

minijackson commented:
I'm interested in this PR. @mkenigs do you have time to fix the merge conflicts?

tfc commented:
I'd really like to see a version of this that really provides two different backends. Factoring out code without having a demonstrator for at least one other backend makes this all very hypothetical. We don't even have to merge it all in one step, but I'd really like to see it work first. Discussing code changes on a demonstrator is better.

minijackson commented:
I think I'm interested in the same kind of backend as @danielfullmer: running tests on physical machines instead of VMs. Unfortunately, for the specific hardware that I'm thinking of, the implementation would be too specific to be in nixpkgs, in my opinion (powerpc64 only, netboot with u-boot, having to upload fpga bitstreams, etc.). I also won't have the time to provide such backend for a while...

tfc commented:
Using the framework for physical hardware is very laborious to set up (i have maintained labs with test HW myself) seems very specific (everyone would have sufficiently different HW so that the software setup is significantly different), that i would even consider using a completely different implementation of the test framework (especially because you can't run such a setup in the nix sandbox) more useful than somehow fiddling all this complexity into this one, resulting in niche-usage anyway. I'd prefer not to mix such immense complexity into this relatively simple tool.

As a good backend we could choose any other hypervisor or container solution. Really anything that could be useful to many.
